### PR TITLE
chore: [IAI-109] Added `redux-saga` effect types through `typed-redux-saga`

### PR DESCRIPTION
**Pull request by dgopsq**
## Short description
This PR is going to add `typed-redux-saga` to the whole codebase.

**The current PR will add an ESLint rule to lint agains the use of the normal `redux-saga/effects`, at the moment the lint rule is just set to a  _warning_.**

An analysis and an explanation about this PR is described [in this document](https://pagopa.atlassian.net/wiki/spaces/IOAPP/pages/466419879/Analisi+utilizzo+di+typed-redux-saga).

## List of changes proposed in this pull request
- Integrated `typed-redux-saga`
- Integrated `typed-redux-saga` Babel macro to re-integrate `redux-saga` at compile time (this will ensure that the transpiled code will be the same as the current one without this PR)
- Integrated an ESLint rule to avoid `redux-saga/effects` import

## How to test
This refactor _shouldn't_ change the current application's behaviour. 

The changes are only on the dev side, adding a type-check to the Redux-Saga effects. Using a `yield* call(() => true)` should automatically type-check the returned type as `Boolean`.

## Help for the reviewer 
Most of the code is actually auto-generated and the overall changes could be summarized in:
1. Used `typed-redux-saga/macro` instead of `redux-saga/effects`
2. Used `yield*` instead of `yield`
3. Moved the `Effect` type from `redux-saga/effecs` to `ReduxSagaEffect`
4. Moved the current return type of a `take([...])` effect inside the `take` generic (i.e. `take<ActionType<...>>([...])`)

### CODE REVIEW

File: ts/sagas/startup/watchLoadMessagesSaga.ts [hunk 1]
```diff
@@ -143,20 +145,28 @@ function* loadMessages(
         // Fetch the services detail in parallel
         // We don't need to store the results because the LOAD_SERVICE_DETAIL_REQUEST is already dispatched by each `loadServiceDetail` action called.
         // We fetch services first because to show messages you need the related service info
-        yield all(
-          pendingServicesIds.map(id => put(loadServiceDetail.request(id)))
+        yield* all(
+          // Here we are not usint the `put` from typed-redux saga
```
Comment (fabriziofff) [inline ts/sagas/startup/watchLoadMessagesSaga.ts:149]:
```suggestion
          // Here we are not using the `put` from typed-redux saga
```

File: ts/sagas/startup/watchLoadMessagesSaga.ts [hunk 2]
```diff
@@ -143,20 +145,28 @@ function* loadMessages(
...
+          pendingServicesIds.map(id => basePut(loadServiceDetail.request(id)))
         );
 
         // Fetch the messages detail in parallel
         // We don't need to store the results because the MESSAGE_LOAD_SUCCESS is already dispatched by each `loadMessage` action called,
         // in this way each message is stored as soon as the detail is fetched and the UI is more reactive.
         // TODO: with the enrichment, we don't need to load each message anymore
-        yield all(pendingMessages.map(_ => put(loadMessageAction.request(_))));
+        yield* all(
+          // Here we are not usint the `put` from typed-redux saga
```
Comment (fabriziofff) [inline ts/sagas/startup/watchLoadMessagesSaga.ts:160]:
```suggestion
          // Here we are not using the `put` from typed-redux saga
```

**Review by fabriziofff — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -172,6 +172,7 @@
     "semver": "^5.7.0",
     "sha.js": "^2.4.11",
     "tslib": "^1.9.3",
+    "typed-redux-saga": "^1.3.1",
```
[the remark this answers is not in the document]
Comment (dgopsq) [reply package.json]: I missed that! It should be done in one of the latest commits :)

File: .eslintrc.js [hunk 4]
```diff
@@ -100,6 +101,14 @@ module.exports = {
       rules: {
         "@typescript-eslint/no-non-null-assertion": "off"
       }
+    },
+    {
+      files: ["./**/*.ts"],
+      excludedFiles: ["./**/*.test.ts"],
+      rules: {
+        "@jambit/typed-redux-saga/use-typed-effects": ["warn", "macro"],
```
[the remark this answers is not in the document]
Comment (dgopsq) [reply .eslintrc.js]: Yep I agree, done!

File: ts/sagas/startup.ts [hunk 5]
```diff
@@ -275,83 +277,88 @@ export function* initializeApplicationSaga(): Generator<Effect, void, any> {
...
     if (!isSessionRefreshed) {
       // The user was previously logged in, so no onboarding is needed
       // The session was valid so the user didn't event had to do a full login,
       // in this case we ask the user to identify using the unlock code.
-      const identificationResult: SagaCallReturnType<
+      // FIXME: This is an unsafe cast caused by a wrongly described type.
+      const identificationResult = (yield* call(
+        startAndReturnIdentificationResult,
+        storedPin
+      )) as unknown as SagaCallReturnType<
```
[the remark this answers is not in the document]
Comment (dgopsq) [reply ts/sagas/startup.ts]: LGTM as a correction, I added it in the latest commits :)

**Review by fabriziofff — APPROVED** [review]

**Pull request merged**